### DISCUSSION TITLE
How to train a image dataset like COCO, which includes json format annotations

### DISCUSSION THREAD

**Discussion by Young-eng** [talmolab/sleap#1620] [Help!]
Hi~ I meet a problem about the training, now I have a image dataset which have been annotated in COCO format, in json file, including train.json and val.json, the dataset is a image dataset, so how can I train it with Sleap ?  Are videos necessary for this workflow with Sleap ? I don't know how to start training using my image dataset. Thanks

**Comment by talmo** [comment]
Hi @Young-eng,

You can import COCO datasets into SLEAP from the GUI by going to **File --> Import**.

Since you have splits, what you could do is then save the SLEAP label to `.slp` files for the training and validation splits separately.

Once you've done that, you can either merge them into a single SLEAP labels file (**File --> Merge**), or if you want to keep the split consistent, you can train via the [`sleap-train`](https://sleap.ai/guides/cli.html#sleap-train) CLI with a syntax like:

```
sleap-train --val labels val.slp training_config.json train.slp
```

But I prefer to actually specify the different splits through the `training_config.json` itself, which has fields for specifying the paths to each split in the data --> labels section:

https://github.com/talmolab/sleap/blob/2d24296c2149402c7ee07601e4d0272fdd308441/sleap/training_profiles/baseline_large_rf.bottomup.json#L4-L5

Let us know if you have any questions!

Cheers,

Talmo

**Reply by Young-eng** [reply]
Thanks for answering!  but I met a new error while importing dataset, it said
  File "D:\Anaconda\envs\sleap\lib\site-packages\sleap\io\format\coco.py", line 78, in read
    skeleton.add_edge(node_names[src_idx], node_names[dst_idx])
  File "D:\Anaconda\envs\sleap\lib\site-packages\sleap\skeleton.py", line 627, in add_edge
    source, destination
ValueError: Skeleton already has an edge between (Right1) and (Right2).

**Reply by talmo** [reply]
Hi @Young-eng,

This is after importing just one of the splits into a new project?

Do you mind sharing the `train.json` file, or at least copying the relevant part that has the info on the skeleton?

Thanks!

Talmo

**Reply by Young-eng** [reply]
I found there were two pairs which were the same in skeleton part in json file, and I deleted one, then it worked. Another problem is about the mermory of GPU, when I took  default hyperparameters for training both centroid and centered instances, it said like "out of memory, .... allocater....." , I forgot to copy error messages. Can I reduce the Max Stride, Input Scale, Output Stride and batch size in both centered and cetroid json files or just in one of them ?

Thanks,

Young

**Reply by Young-eng** [reply]
And when I trained centered instances, it got stuck with no progress for minutes, it took long time for each epoch, and it crashed once, here are some messages:

I tensorflow/stream_executor/cuda/cuda_blas.cc:1760] TensorFloat-32 will be used for the matrix multiplication. This will only be logged once.
 W tensorflow/core/grappler/costs/op_level_cost_estimator.cc:690] Error in PredictCost() for the op: op: "CropAndResize" attr { key: "T" value { type: DT_FLOAT } } attr { key: "extrapolation_value" value { f: 0 } } attr { key: "method" value { s: "bilinear" } } inputs { dtype: DT_FLOAT shape { dim { size: 26 } dim { size: 678 } dim { size: 678 } dim { size: 1 } } } inputs { dtype: DT_FLOAT shape { dim { size: -2 } dim { size: 4 } } } inputs { dtype: DT_INT32 shape { dim { size: -2 } } } inputs { dtype: DT_INT32 shape { dim { size: 2 } } } device { type: "GPU" vendor: "NVIDIA" model: "GeForce RTX 3090" frequency: 1695 num_cores: 82 environment { key: "architecture" value: "8.6" } environment { key: "cuda" value: "11020" } environment { key: "cudnn" value: "8100" } num_registers: 65536 l1_cache_size: 24576 l2_cache_size: 6291456 shared_memory_size_per_multiprocessor: 102400 memory_size: 23402708992 bandwidth: 936096000 } outputs { dtype: DT_FLOAT shape { dim { size: -2 } dim { size: -5 } dim { size: -6 } dim { size: 1 } } }
 W tensorflow/core/grappler/costs/op_level_cost_estimator.cc:690] Error in PredictCost() for the op: op: "CropAndResize" attr { key: "T" value { type: DT_FLOAT } } attr { key: "extrapolation_value" value { f: 0 } } attr { key: "method" value { s: "bilinear" } } inputs { dtype: DT_FLOAT shape { dim { size: 1 } dim { size: 3500 } dim { size: 2160 } dim { size: 3 } } } inputs { dtype: DT_FLOAT shape { dim { size: -2 } dim { size: 4 } } } inputs { dtype: DT_INT32 shape { dim { size: -2 } } } inputs { dtype: DT_INT32 shape { dim { size: 2 } } } device { type: "CPU" vendor: "GenuineIntel" model: "101" frequency: 2200 num_cores: 10 environment { key: "cpu_instruction_set" value: "AVX SSE, SSE2, SSE3, SSSE3, SSE4.1, SSE4.2" } environment { key: "eigen" value: "3.3.90" } l1_cache_size: 32768 l2_cache_size: 1048576 l3_cache_size: 14417920 memory_size: 268435456 } outputs { dtype: DT_FLOAT shape { dim { size: -2 } dim { size: 2712 } dim { size: 2712 } dim { size: 3 } } }

 W tensorflow/core/grappler/costs/op_level_cost_estimator.cc:690] Error in PredictCost() for the op: op: "CropAndResize" attr { key: "T" value { type: DT_FLOAT } } attr { key: "extrapolation_value" value { f: 0 } } attr { key: "method" value { s: "bilinear" } } inputs { dtype: DT_FLOAT shape { dim { size: 1 } dim { size: 3500 } dim { size: 2160 } dim { size: 3 } } } inputs { dtype: DT_FLOAT shape { dim { size: -2 } dim { size: 4 } } } inputs { dtype: DT_INT32 shape { dim { size: -2 } } } inputs { dtype: DT_INT32 shape { dim { size: 2 } } } device { type: "CPU" vendor: "GenuineIntel" model: "101" frequency: 2200 num_cores: 10 environment { key: "cpu_instruction_set" value: "AVX SSE, SSE2, SSE3, SSSE3, SSE4.1, SSE4.2" } environment { key: "eigen" value: "3.3.90" } l1_cache_size: 32768 l2_cache_size: 1048576 l3_cache_size: 14417920 memory_size: 268435456 } outputs { dtype: DT_FLOAT shape { dim { size: -2 } dim { size: 2712 } dim { size: 2712 } dim { size: 3 } } }
/var/spool/slurm/d/job10434/slurm_script: line 59:   406 Killed                  sleap-train --val_labels  val_labels.v001.pkg.slp centered_instance.json train_labels.v001.pkg.slp, It crashed again. 


Thanks !

**Reply by talmo** [reply]
> I found there were two pairs which were the same in skeleton part in json file, and I deleted one, then it worked. Another problem is about the mermory of GPU, when I took default hyperparameters for training both centroid and centered instances, it said like "out of memory, .... allocater....." , I forgot to copy error messages. Can I reduce the Max Stride, Input Scale, Output Stride and batch size in both centered and cetroid json files or just in one of them ?
> 
> Thanks,
> 
> Young

Hi @Young-eng,

Right, so the next thing you may want to try is to disable data preloading. This is an optimization for training which works quite well at speeding things up by loading all of your images into memory, thereby avoiding the I/O bottleneck of loading the images repeatedly during training.

The trade-off is that it's pretty memory intensive, so it might not work well on larger datasets, larger image sizes, or if you have a system with less memory capacity.

Based on your description that it takes a while for the training to start and then leading to OOM errors, I'm guessing that's what's happening. You can test it out by disabling the preloading in your training job config file, namely this part of the optimization section:

https://github.com/talmolab/sleap/blob/2d24296c2149402c7ee07601e4d0272fdd308441/sleap/training_profiles/baseline_large_rf.bottomup.json#L61-L62

Give that a go and let us know if you're still running into issues!

Cheers,

Talmo

PS: The next thing to try if you want more control over the training configuration, would be to run your own [interactive training](https://sleap.ai/notebooks/Interactive_and_resumable_training.html).

**Reply by Young-eng** [reply]
Hi Talmo!
Thanks for your kind replies!

I need to pick up the Sleap again recently, and I met some problems that I am not very clear. It is about the training and validation dataset.

Since I imported COCO train.json and val.json to get train.pkg.slp and val.pkg.slp. I changed to path centered.json and centroid.json in training.job. like this one, 
![image](https://github.com/talmolab/sleap/assets/55878284/7a6b3d94-406e-4407-8aa1-e35f38bba08e)

while the default value of training_labels and validation_labels are null. And then started training. I got 0.09 mAP for train, however got 0.0 mAP for val. I am wandering about the 0.0. mAP and if any steps are wrong above.  I used following code snippets to get metrics.
`metrics = sleap.load_metrics("/models_v001/240601_003535.centered_instance/", split="val")`

Another thing is, that at first, I did not export val.pkg.slp and just used train.pkg.slp for training, and did not set any training_labels or validation_labels in centered/centroid.json files, and keep default values(Null). While training was done, in the models there were val.slp and val.npz, and I also evaluate the model with the code:

`metrics = sleap.load_metrics("/models_v001/240601_003535.centered_instance/", split="val")`

and also got some values(e.g. mAP, mAR) different from 

`metrics = sleap.load_metrics("/models_v001/240601_003535.centered_instance/", split="train")`

, so I wander how it get the validation data in training process while I just use train.pkg.slp. Does it split the training.pkg.slp into two parts, train and val during training? If it is the case, what is the ratio between the new train and val dataset ? 

Thanks again,

Young

**Reply by eberrigan** [reply]
Do you mind uploading your training configs [here](https://docs.google.com/forms/d/e/1FAIpQLSfWBwcuZH8lxGyFabMAODLjSUE4YrSCYbrrnYvXCpgBqZpcJA/viewform)? Then we can give you a better answer. Thanks!

**Reply by Young-eng** [reply]
I have uploaded the training configs including centroid and centered config. I think I know the answer about the split of train and val data when no validation path is assigned. There is a parameter named validation fraction. val data would be sampled from train data in a given proportion. 

Another problem I met was crush when the input image size was large, e.g $1900 \times 3500$, even though I set the batch size to 1, adjust the input scale and filter_rate, etc. The usage of RAM is over 100G while the memory of GPU is surplus and not full utilized when training centered model.  Are there any methods which can resize the input images to a small size for training? 

Thanks.

**Reply by talmo** [reply]
Hi @Young-eng,

You can also disable data preloading if you have a large dataset:

https://github.com/talmolab/sleap/blob/834c68a11842b843d3c191041f5b2390eae307ce/sleap/training_profiles/baseline_large_rf.bottomup.json#L62

This normally will speed up the data loading pipeline during training quite a bit, but it'll blow up your RAM if your data/images are too large. Try turning that off and seeing if it helps with the memory usage.

Cheers,

Talmo

**Reply by Young-eng** [reply]
I did that, I  tried to set "preload_data" to false. But It still crushed at the last stage during the training of centered model. 

Thanks so much!

**Reply by talmo** [reply]
Hmmm, strange! Are you sure those are set in the centered instance model configuration in addition to the centroid model one?

A couple more settings to try:

```
        "online_shuffling": true,
        "shuffle_buffer_size": 4,
        "prefetch": false,
```

The shuffling preloads a lot of images into a buffer for order randomization, but it doesn't matter as much if you have a lot of images. Decreasing the buffer size helps, but you might be able to just turn it off entirely.

You can also turn off prefetching. I'm not sure about the impact on RAM, but it might help a bit too since it'll disable more preloading optimizations.

Cheers,

Talmo

**Reply by Young-eng** [reply]
Here are parameters in centered model:
![image](https://github.com/talmolab/sleap/assets/55878284/57cf9931-8834-48b7-baeb-e412f2742454)

It failed to compute metrics. Error like this, but I need the metric. I did not encounter this problem when training other datasets with smaller image size. It is strange :(
![image](https://github.com/talmolab/sleap/assets/55878284/304fd917-cc9e-437e-aa5a-791d2b3c62c8)

**Comment by Young-eng** [comment]
I solved this problem by resizing original datasets to smaller size, then import transformed datasets for training.
